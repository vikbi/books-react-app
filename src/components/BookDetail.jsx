import React from 'react'

const BookDetail = ({book, toggleSelection}) => {

    const toggleCheck = (status) => {
        toggleSelection(book.id, status)
      }
    
      return (
        <div key={book.id} className="book-detail" data-testid="book-detail">
          <input key={book.id} checked={book.is_checked} type="checkbox" onChange={(e) => toggleCheck(e.target.checked)}/>
          <img className="book-logo" src={book.image} data-testid="book-logo"/>
          <div className="book-header">
            <label data-testid="book-title">{book.title}</label>
            <div>{book.authors.map((a,i) => <span data-testid="book-author" key={i}>{a}</span>)}</div>
          </div>
        </div>
      )
}

export default BookDetail
