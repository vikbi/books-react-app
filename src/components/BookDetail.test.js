import BookDetail from './BookDetail';
import userEvent from '@testing-library/user-event';
import { cleanup, render, screen, act } from '@testing-library/react';
import renderer from "react-test-renderer";

const mockBook = {
    id: '1',
    title: 'test book',
    image: 'image-link',
    authors: ['author1', 'author2'],
    is_checked: false
}
const toggleFn = jest.fn();

describe('Book Detail tests', () => {

  afterEach(cleanup);

  it('Should render Book Details correctly', () => {
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />)
    const element = screen.getByTestId('book-detail');
    expect(element).toBeInTheDocument();
  });

  it('Should toggleSelection on checkbox : checked', async () => {
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />);
    const checkboxElement = screen.getByRole("checkbox");
    await userEvent.click(checkboxElement);
    
    expect(toggleFn).toHaveBeenCalledWith(mockBook.id,true);
  })

  it('Should toggleSelection on checkbox : unchecked',async () => {
    const mockBook = {
        id: '1',
        title: 'test book',
        image: 'image-link',
        authors: ['author1', 'author2'],
        is_checked: true
    }
    
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />);
    const checkboxElement = screen.getByRole("checkbox");

    await userEvent.click(checkboxElement);
    expect(toggleFn).toHaveBeenCalledWith(mockBook.id,false);
  })

  it('Should list authors correctly', () => {
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />);
    const authorList = screen.getAllByTestId('book-author');
    expect(authorList.length).toEqual(2)
  })

  it('Should have the correct book title', async () => {
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />);
    expect(await screen.getByText(mockBook.title)).toBeTruthy();
  });

  it('Should have default checkbox unchecked for given book', async () => {
    render(<BookDetail book={mockBook} toggleSelection={toggleFn} />);
    expect(await screen.getByRole('checkbox')).toBeTruthy();
    expect(await screen.getByRole('checkbox')).not.toBeChecked();
  });

  it("Matches DOM Snapshot", () => {
    const domTree = renderer.create(<BookDetail book={mockBook} toggleSelection={toggleFn} />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});