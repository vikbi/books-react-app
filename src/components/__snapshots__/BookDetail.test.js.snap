// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Detail tests Matches DOM Snapshot 1`] = `
<div
  className="book-detail"
  data-testid="book-detail"
>
  <input
    checked={false}
    data-testid="book-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <img
    className="book-logo"
    data-testid="book-logo"
    src="image-link"
  />
  <div
    className="book-header"
  >
    <label
      data-testid="book-title"
    >
      test book
    </label>
    <div>
      <span
        data-testid="book-author"
      >
        author1
      </span>
      <span
        data-testid="book-author"
      >
        author2
      </span>
    </div>
  </div>
</div>
`;
